#149955236 Add updated_at column to contribution model

wanjiku-carol opened the pull request:
**What does this PR do ?**


- Adds an updated_at column to the Contribution model

**Description of Task to be completed ?**


- Adding a updated_at column to the Contributions model. 


> updated_at = db.Column(db.DateTime, default=datetime.utcnow(), onupdate=datetime.utcnow())


- Running migrations to incorporate changes made


**How should this be manually tested?**


- Run `pytest` or `pytest -v` for `def test_updated_at` in `test_models.py`


**What are the relevant pivotal tracker stories?**


#149955236

njirap commented on test/api/test_models.py:
```diff
@@ -170,3 +170,20 @@ def test_contribution_type_relation(self):
+            )
+        db.session.add(contrib)
+        db.session.commit()
+        db.session.query(Contribution).filter(Contribution.id==contrib_id).update({'title': contrib_title})
```
why not `Contribution.query.filter`?

njirap commented on test/api/test_models.py:
```diff
@@ -170,3 +170,20 @@ def test_contribution_type_relation(self):
+        db.session.commit()
+        retrieve_updated_contrib = Contribution.query.filter_by(id=contrib_id).first()
+        self.assertIsNotNone(retrieve_updated_contrib.updated_at)
+        self.assertIsInstance(retrieve_updated_contrib.updated_at, datetime)
```
lets start by asserting that on creation, the `created_at` and `updated_at` are equal, then update, then assert that the two are not equal

wanjiku-carol commented:
@njirap I have added `self.assertEqual` and  `self.assertNotEqual`tests. However, I had to change the formats because of the differences in milliseconds.

I have also adjusted the format of this `Contribution.query.filter(Contribution.id==contrib_id).update({'title': contrib_title})`

njirap commented on test/api/test_models.py:
```diff
@@ -170,3 +170,24 @@ def test_contribution_type_relation(self):
+        new_contrib = Contribution.query.filter_by(id=contrib_id).first()
+        self.assertEqual(new_contrib.created_at.strftime('%B %d %Y - %H:%M:%S'), new_contrib.updated_at.strftime('%B %d %Y - %H:%M:%S'))
+
+        Contribution.query.filter(Contribution.id==contrib_id).update({'title': contrib_title})
```
blank lines too many here

njirap replied:
refactor this too

`Contribution.query.filter_by(id=contrib_id).first()`

wanjiku-carol replied:
@njirap On it

njirap replied:
did it work?

wanjiku-carol replied:
This worked.Let me push the code.

njirap commented on test/api/test_models.py:
```diff
@@ -170,3 +170,24 @@ def test_contribution_type_relation(self):
+        updated_contrib = Contribution.query.filter_by(id=contrib_id).first()
+        self.assertIsNotNone(updated_contrib.updated_at)
+        self.assertIsInstance(updated_contrib.updated_at, datetime)
+        self.assertNotEqual(new_contrib.created_at, new_contrib.updated_at)
```
Also here assert on the `.strftime('%B %d %Y - %H:%M:%S')`

wanjiku-carol replied:
@njirap, When I use this format, the test fails because the created_at and updated_at are equal.  They only vary with milliseconds hence the current format is the only way the test passes. The same applies for the test before updating. I presume this is because each line of code is executed step by step hence the created_at and updated_at are recorded in the model one after the other, and they may always vary by milliseconds as long as they occur within the same function. Just a theory.

njirap replied:
very smart Proffesor Shikoh. I concur.

njirap commented:
LGTM
